### DISCUSSION TITLE
Need support with pipes (come from akka.net)

### DISCUSSION THREAD

**Discussion by LaszloLueck** [louthy/language-ext#1157] [Q&A] [answered]
Hi,
as described in the title i need help converting code in akka.net to LanguageExt pipes.
I could continue to use akka.net, but I have LanguageExt in use anyway, so I don't need another library if LanguageExt can do the job.

That's my implementation with akka.net
The example is very finely structured. Normally, individual steps would be summarised. This is only about the comparison.

```
(01) await Source
(02)    .From(Directory.GetFiles(config.WriteTo, "*.json", SearchOption.AllDirectories))
(03)    .SelectAsync(5, async filePath => await File.ReadAllTextAsync(filePath))
(04)    .Where(content => content.Any())
(05)    .Select(content => Encoding.UTF8.GetBytes(content))
(06)    .Select(bytes => new MemoryStream(bytes))
(07)    .SelectAsync(5, async stream => (await JsonSerializer.DeserializeAsync<Recording>(stream)).ToOpt())
(08)    .GroupedWithin(1000, TimeSpan.FromSeconds(5))
(09)    .Select(group => group.Somes())
(10)    .Select(GenerateElasticRecords)
(11)    .Select(group =>
        {
            logger.LogInformation("process {Group} elements to index", group.Length());
            return group;
        })
(12)    .SelectAsync(2, async group =>
        {
            await elasticHelper.BulkIndexDocuments(esClient, config.IndexName, group.ToArray());
            return unit;
        })
(13)    .RunWith(Sink.Ignore<Unit>(), _actorSystem.Materializer());
```

(01) and (02): Build a source from files in a directory, corresponds to LanguageExt Producer
(03): my first question. I need a corresponding thing in LanguageExt, something like PipeAsync? Reads async a file from path
(04): equivalent to LanguageExt Filter, if file content is empty, filter that away
(05): equivalent to LanguageExt Pipe. Encode the content to utf8 byte array
(06): equivalent to LanguageExt Pipe. Creates a memorystream from byte array
(07): ???. Creates a concrete Object from json
(08): ???. Groupes single elements to a bunch of things (1000) or all elements within 5 seconds
(09): equivalent to LanguageExt Pipe. removes the Nones from List
(10): equivalent to LanguageExt Pipe. Method converts objects to transport objects for an elastic index
(11): equivalent to LanguageExt Pipe. Only for logging
(12): ???. Bulk insert the grouped elements to an elastic instance
(13): Is corresponds this to a Consumer? The _actorSystem.Materializer ist the LanguageExt Runtime? Think so.

There are, in some other projects, merge / broadcast stages and so on.
In addition to my questions above, I would like to know: Can or should one compare the two libraries at all?
What would the (exemplary) control logic in LanguageExt look like?
Does LanguageExt know something like backpressuring (when you work with fast producer and slow consumer) and parallelism how the first number can be specified in the SelectAsync.

Thanks in Advance,
Regards,
Laszlo

**Comment by louthy** [comment, marked as answer]
I think in this case Pipes probably isn't what you need, mostly because it's more like the linux pipe operator.  It can certainly do what you need, but it would involve building a signalling system for the time-based-yielding and flushing-what's-left-in-an-aggregator when the stream is complete.  Presumably those capabilities are built-in to Akka streams (I haven't used them, so I don't know what it's underlying architecture is).

The new Transducer based streams that I'm building into all language-ext monads is exactly what you need - and are much easier to use than Pipes, but I am super busy with other projects at the moment, so I won't be releasing that for quite a while.

Other options would be using Rx.NET.  You can lift Observables into Pipes if you wanted to.

**Comment by LaszloLueck** [comment]
Thank you @louthy for your explanation.
In the meantime I continue to use akka.net, I do use that in more projects than I initially thought.

Regards,
Laszlo
